# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Allocator.cpp
# Commit: 43978ebfc7ac
# Full Hash: 43978ebfc7ace2c37aba0b7644ac4e866e0730bd
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2021-01-20 16:13:57
# Regressor Bug: 1686219
# File Overlap Count: 1
# Description:
#   Bug 1686219 - Store chunk information at the start rather than the end of the chunk r=sfink
#   
#   This moves the chunk metadata to the start of the chunk and defines the data structures in the public header.  This simplifies accessing this data and removes the need for hardcoded offsets.
#   
#   Requesting review from jandem for JIT updates.
# ==============================================================================

diff -r 496d47ff3fa1 -r 43978ebfc7ac js/src/gc/Allocator.cpp
--- a/js/src/gc/Allocator.cpp	Wed Jan 20 10:04:11 2021 +0000
+++ b/js/src/gc/Allocator.cpp	Wed Jan 20 10:04:29 2021 +0000
@@ -788,7 +788,7 @@
 }
 
 void GCRuntime::recycleChunk(Chunk* chunk, const AutoLockGC& lock) {
-  AlwaysPoison(&chunk->trailer, JS_FREED_CHUNK_PATTERN, sizeof(ChunkTrailer),
+  AlwaysPoison(&chunk->header, JS_FREED_CHUNK_PATTERN, sizeof(ChunkHeader),
                MemCheckKind::MakeNoAccess);
   emptyChunks(lock).push(chunk);
 }
@@ -861,6 +861,9 @@
   Poison(this, JS_FRESH_TENURED_PATTERN, ChunkSize,
          MemCheckKind::MakeUndefined);
 
+  new (&header) ChunkHeader(gc->rt);
+  info.init();
+
   /*
    * We clear the bitmap to guard against JS::GCThingIsMarkedGray being called
    * on uninitialized data, which would happen before the first GC cycle.
@@ -873,10 +876,6 @@
    */
   decommitAllArenas();
 
-  /* Initialize the chunk info. */
-  info.init();
-  new (&trailer) ChunkTrailer(gc->rt);
-
   /* The rest of info fields are initialized in pickChunk. */
 }
 